File: plugins/grok-augment/scripts/clean_test.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import stat
import subprocess
import sys
import tempfile
from pathlib import Path


PLUGIN_ROOT = Path(__file__).resolve().parents[1]
REPO_ROOT = PLUGIN_ROOT.parents[1]


def run(args: list[str], env: dict[str, str] | None = None) -> None:
    completed = subprocess.run(
        args,
        cwd=REPO_ROOT,
        env=env,
        text=True,
        capture_output=True,
        check=False,
    )
    if completed.returncode != 0:
        sys.stderr.write(completed.stdout)
        sys.stderr.write(completed.stderr)
        raise SystemExit(completed.returncode)
    sys.stdout.write(completed.stdout)
    sys.stderr.write(completed.stderr)


def main() -> int:
    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp)
        fake_home = root / "home"
        fake_codex_home = root / "codex"
        fake_bin = root / "bin"
        fake_home.mkdir()
        fake_codex_home.mkdir()
        fake_bin.mkdir()
        fake_grok = fake_bin / "grok"
        fake_grok.write_text(
            "#!/bin/sh\n"
            "case \"$*\" in\n"
            "  *--version*) echo 'grok 0.0.clean-test'; exit 0 ;;\n"
            "  *models*) echo 'grok-build'; exit 0 ;;\n"
            "  *) echo 'CLEAN_TEST_GROK_RESPONSE'; exit 0 ;;\n"
            "esac\n",
            encoding="utf-8",
        )
        fake_grok.chmod(fake_grok.stat().st_mode | stat.S_IXUSR)
        env = os.environ.copy()
        env.update(
            {
                "HOME": str(fake_home),
                "CODEX_HOME": str(fake_codex_home),
                "GROK_AUGMENT_GROK_BIN": str(fake_grok),
            }
        )
        run(["python3", "-m", "unittest", "discover", "-s", "plugins/grok-augment/tests"], env)
        run(["python3", str(PLUGIN_ROOT / "scripts/grok_augment.py"), "inspect", "--json"], env)
        run(
            [
                "python3",
                str(PLUGIN_ROOT / "scripts/grok_augment.py"),
                "--json",
                "video",
                "one safe closed-loop test shot",
            ],
            env,
        )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/grok-augment/scripts/clean_test.ts
#!/usr/bin/env -S node --experimental-strip-types
import { spawnSync } from "node:child_process";
import assert from "node:assert/strict";
import { chmodSync, mkdirSync, mkdtempSync, readdirSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

const PLUGIN_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const REPO_ROOT = path.resolve(PLUGIN_ROOT, "../..");

function run(args: string[], env?: NodeJS.ProcessEnv): void {
  const [command, ...rest] = args;
  const completed = spawnSync(command, rest, {
    cwd: REPO_ROOT,
    env,
    encoding: "utf8",
  });
  process.stdout.write(completed.stdout || "");
  process.stderr.write(completed.stderr || "");
  if (completed.status !== 0) process.exit(completed.status || 1);
}

function validateNoPythonFiles(root: string): void {
  const stack = [root];
  while (stack.length) {
    const dir = stack.pop()!;
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      if ([".git", "node_modules"].includes(entry.name)) continue;
      const file = path.join(dir, entry.name);
      if (entry.isDirectory()) stack.push(file);
      else assert.ok(!file.endsWith(".py"), `unexpected Python file: ${path.relative(root, file)}`);
    }
  }
}

function main(): number {
  validateNoPythonFiles(PLUGIN_ROOT);
  const root = mkdtempSync(path.join(tmpdir(), "grok-augment-clean-"));
  const fakeHome = path.join(root, "home");
  const fakeCodexHome = path.join(root, "codex");
  const fakeBin = path.join(root, "bin");
  mkdirSync(fakeHome);
  mkdirSync(fakeCodexHome);
  mkdirSync(fakeBin);
  const fakeGrok = path.join(fakeBin, "grok");
  writeFileSync(
    fakeGrok,
    `#!/bin/sh
case "$*" in
  *--version*) echo 'grok 0.0.clean-test'; exit 0 ;;
  *models*) echo 'grok-build'; exit 0 ;;
  *) echo 'CLEAN_TEST_GROK_RESPONSE'; exit 0 ;;
esac
`,
    "utf8",
  );
  chmodSync(fakeGrok, 0o755);
  const env = {
    ...process.env,
    HOME: fakeHome,
    CODEX_HOME: fakeCodexHome,
    GROK_AUGMENT_GROK_BIN: fakeGrok,
  };
  run(["node", "--experimental-strip-types", "--test", "plugins/grok-augment/tests/*.ts"], env);
  run(["node", "--experimental-strip-types", path.join(PLUGIN_ROOT, "scripts/grok_augment.ts"), "inspect", "--json"], env);
  run(
    [
      "node",
      "--experimental-strip-types",
      path.join(PLUGIN_ROOT, "scripts/grok_augment.ts"),
      "--json",
      "video",
      "one safe closed-loop test shot",
    ],
    env,
  );
  return 0;
}

process.exitCode = main();
